v2/devdoc/logging.md

```diff
@@ -0,0 +1,162 @@
+    LOG_CONTEXT_CREATE(context_2, context_1, LOG_CONTEXT_FIELD("last name", "%s", "uaga"), LOG_CONTEXT_FIELD("age", "%d", 42));
+
+    logger_log(LOG_LEVEL_VERBOSE, context_1, "log context 1 data");
+    logger_log(LOG_LEVEL_VERBOSE, context_2, "log context 2 data");
+
+    LOG_CONTEXT_DESTROY(context_1);
+
+    // stack allocated context
+    LOCAL_LOG_CONTEXT_DEFINE(local_context, context_2, LOG_CONTEXT_FIELD("the knights that say", "%s", "Nee!"));
+
+    logger_log(LOG_LEVEL_VERBOSE, &local_context, "log with local context");
+
+    LOG_CONTEXT_DESTROY(context_2);
+```
+
+## Log levels
+
+The following log levels shall be exposed by the library:
+
+- `CRITICAL`
+
+- `WARNING`
+
+- `ERROR`
+
+- `INFO`
+
+- `VERBOSE`
+
+## Sinks
+
+Multiple sinks shall be supported. Setting up globally which sinks to be used shall be supported (so that they do not have to be specified every time a `c_logging_log` call is made).
+
+Defining the sinks to use shall be done by defining a NULL terminated array, where each entry shall be a pointer to a structure that represents the sink interface.
+
+Example:
+
+```c
+const LOG_SINK* log_sinks[] = { &console_log_sink, &etw_log_sink, NULL };
+```
+
+### ETW sink
+
+An `ETW` sink shall be supported so that events are emitted via TraceLogging.
+
+For the `ETW` sink there should be a posibility of specifying the provider to use.
+
+### console sink
+
+A console sink shall be supported so that events are printed to the console using `printf`.
+
+## API
+
+```c
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* format, ...);
+```
+
+`log_context` may be `NULL`. If `log_context` is `NULL`, no context information shall be added to the log line.
+
+`format` shall be a `printf`-like format string.
+
+`...` shall be the variable number of arguments required by `format`.
+
+## Contextual logging
+
+For each log line the library shall support specifying a context to be added to the log line.
+
+The context shall allow defining pairs of properties that are added to the event when the event is emitted.
+
+### Local stack context creation
+
+The library shall support defining a context using only stack memory, not requiring any memory allocation:
+
+```c
+#define LOCAL_LOG_CONTEXT_DEFINE(log_context, parent_context, ...) \
+```
+
+where `...` is a list of context fields defined using `LOG_CONTEXT_FIELD`.
+
+Example:
+
+```c
+    LOCAL_LOG_CONTEXT_DEFINE(log_context, NULL, LOG_CONTEXT_FIELD("property_name", "%s", MU_P_OR_NULL(prop_value)));
+```
+
+Bonus:
+
+An ideal implementation for the stack allocated contexts is also to not do any memory copy until actually needed (until they are used).
+
+That would allow wring code like below while not performing the memory copies until the logging is actially executed.
+
+```c
+void a(LOG_CONTEXT_HANDLE log_context, const char* correlation_id)
+{
+    LOCAL_LOG_CONTEXT_DEFINE(local_context, NULL, LOG_CONTEXT_FIELD("correlation_id", "%s", correlation_id));
```

Comment:
>correlation_id [](http://example.com/codeflow?start=92&length=14)

when is "correlation_id" evaluated? At context creation time or at every logger_log?

One would "suppose" it is at every logger-log, otherwise, it being static, can just be a string (unless we care about data structure in the string).

But on the other hand, it is rather impossible to "store" things from one macro to the other, so... then it is statically evaluated.

Are these fields going to be added to the tracelogging structure and ... have some dynamic tracelogging?

Reply:
It is evaluated at context definition/allocation time.
I think it would be somewhat misleading to evaluate it at logging time.